### ISSUE TITLE
Degraded connection rate/timeouts with 2.6.5

### ISSUE THREAD

**Issue by rnsanchez** [haproxy/haproxy#1895]
### Detailed Description of the Problem

We have noticed a degraded connection rate, leading to severe throughput issues, on a mildly-loaded server (25~30 Gbps).
Upon realizing that HAProxy-2.6.5 was in a degraded state, we forcefully restarted it (`killall -KILL haproxy; /etc/rc.d/rc.haproxy start`) because we needed the new instance to instantly listen on `:80`.

At this time, we are not aware of conditions that might lead to this behavior.  An initial set of this condition _seems_ to be related to an orderly restart (i.e., throughout the day, changes to the config might trigger an `-sf` reload).  However, we witnessed this degradation happen successively, very close to cold restarts, and minutes apart.  With 6 restarts in a 16-minute window, it was decided a roll-back to 2.4.18 was in order.

It was possible to gather some information from one of the early restarts (most likely; it was imperative to restore service availability).

As an example from our monitoring, there was no unexpected load or system behavior; HAProxy seemed to simply choke, being extremely slow to respond to a `telnet localhost 80` (many seconds).

[activity.log](https://github.com/haproxy/haproxy/files/9781616/activity.log)
[fd.log](https://github.com/haproxy/haproxy/files/9781617/fd.log)

10-second profiling:
```
Tasks activity:
  function                      calls   cpu_tot   cpu_avg   lat_tot   lat_avg
  sc_conn_io_cb               1063910   5.291s    4.973us      -         -   
  h1_io_cb                     518541   218.3ms   421.0ns      -         -   
  process_stream                 5339   132.5ms   24.82us   7.352ms   1.376us
  h2_io_cb                       1682   45.64ms   27.13us      -         -   
  h1_timeout_task                 572   1.244us   2.000ns   8.701ms   15.21us
  accept_queue_process            326   2.832ms   8.687us      -         -   
  ssl_sock_io_cb                  104   4.537ms   43.62us      -         -   
  task_run_applet                  47   655.1us   13.94us   146.5us   3.116us
  main+0xfd030                     15   15.17us   1.011us   19.19us   1.279us
  h2_timeout_task                   4      -         -      270.8us   67.70us
  srv_cleanup_toremove_conns        4   67.69us   16.92us   90.86us   22.71us
  srv_cleanup_idle_conns            3   26.73us   8.909us   3.735us   1.245us
```

In the third pane, it is easy to tell by the sudden drops at what points in time we forcefully restarted HAProxy (note: network interface is degraded; it was expected to operate at 40+ Gbps):
![Screenshot 2022-10-13 at 21-51-40 Grafana - HAProxy](https://user-images.githubusercontent.com/87608/195739711-5e8a0ea4-eadd-459c-8f40-ff647b2baae2.png)

![Screenshot 2022-10-13 at 22-17-39 Grafana - Hyper](https://user-images.githubusercontent.com/87608/195740121-75ebd3f8-71a5-437f-83eb-b5cd11dd5841.png)

### Expected Behavior

Well, it should not degrade.  :-)

### Steps to Reproduce the Behavior

We don't know what conditions are required to enter this state, nor how to reproduce.  To detect, it seems that a timeout detection could give early signs of service degradation.

### What is your configuration?

```haproxy
global
    nbthread 35
    hard-stop-after 7d
    maxconn 900000
    ulimit-n 3150919
    user haproxy
    group haproxy
    daemon
    unix-bind user haproxy mode 777

    node haproxy-frontend

    tune.ssl.default-dh-param 2048
    tune.ssl.maxrecord 1419

    tune.ssl.cachesize 100000
    tune.ssl.lifetime 600

    tune.idletimer 1000
    tune.bufsize 65536 # Para absorver melhor dados do varnish.

    stats socket /var/run/haproxy_conexoes.sock mode 666 level admin expose-fd listeners
    stats timeout 10m

    log /var/tmp/logtape.unix local0

    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256

    ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA
    ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256

    ssl-default-bind-options no-tls-tickets
    ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

defaults
    mode    http

    option abortonclose

    option splice-response
    retries 1
    maxconn 900000
    timeout connect 1s
    timeout server 100s
    timeout server-fin 600ms
    timeout client 100s
    timeout client-fin 600ms
    timeout http-request 6s
    timeout http-keep-alive 40s
    timeout tunnel 300s

frontend ssl
    log global

    # binds..

    # acls..

    # captures..

    # request adjustments..

    use_backend ..
    default_backend ..

backend ..
    http-reuse always
    http-response set-header access-control-allow-origin "*"
    http-request set-header cf-connecting-ip ..

    timeout http-keep-alive 1s
    server v01 abns@..

backend ..
    http-reuse never
    http-response set-header access-control-allow-origin "*"
    http-request set-header cf-connecting-ip ..
    http-response set-header Date %[date(),http_date]

    server v01 abns@..
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.5-987a4e2 2022/09/03 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.5.html
Running on: Linux 5.15.23 #2 SMP PREEMPT Sat Feb 12 21:05:18 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_NS=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=64).
Built with OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
Running on OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Additional Information

It could be somehow, but not exclusively, triggered after a `-sf` reload.  But this is a hunch.

**Comment by rnsanchez** [comment]
Also confirmed on 2.4.17.  Apologies, I had included `-vv` for 2.4.18, and edited this comment.

**Comment by capflam** [comment]
So it means the issue is also present on 2.4.17 but not on 2.4.18 ?

**Comment by rnsanchez** [comment]
> So it means the issue is also present on 2.4.17 but not on 2.4.18 ?

Sorry, let me rephrase: we have confirmed this behavior happens on 2.6.5 and 2.4.17.  We're now using 2.4.18 and, while at it tracking whether it will happen as well.

Our HAProxy instance is restarted (gracefully) many times throughout 24h, but we don't know yet what's needed to trigger this particular issue.  We'll try a different way of collecting data, in the case we hit this again---when it happens, we can't get any more data from HAProxy (i.e., we lose telemetry), and so far this is one of the ways to confirm the issue happened (other than a brief service outage until HAProxy is hard-restarted).

**Comment by capflam** [comment]
Ok, copy that !

**Comment by felipewd** [comment]
@capflam This is really tricky to catch in production...since when we do there's an almost total collapse of traffic, either with the `bind` socket throwing `connection refused` or with it being way way slow (`telnet localhost 80` taking 2-3 *seconds* to accept the connection), we have to kill haproxy and restart.

BUT we've seen it with both 2.4 and 2.6 for sure (2.4.17, 2.4.18 _once_, 2.6.4, 2.6.5, and 2.6.6 also once). Since there are no relevant (for our setup) bug fixes on 2.4.19 for our needs, we haven't updated them all.

The data here is from 2.6.5, when it happened last week and frankly we were able to keep our cool long enough and don't freak out so fast :-)

**Comment by felipewd** [comment]
And just to be clear: The frontend symptom when the data on this ticket was collected was a local telnet to port 80 took between 1-4 seconds to accept. The Currconn dropped almost to all and the machine itself didn't have any indication of any problem. We're not sure if either fd.log or activity.log helps tracking what kind of problem this could be.

**Comment by rnsanchez** [comment]
In the most peculiar way, we have an odd-ball development.

So here's what we have: we open a custom-crafted monitoring agent, which connects to the stats socket a first time, leaving that connection idle.  Every few seconds, a second connection is attempted and, if that fails, a bunch of stats commands would be sent over the first (connected) socket (at least that is what we wished).  And every so in a while, the first connection has to be remade, e.g.: stats socket timeout.

Basically:
```
    stats socket /var/run/haproxy.sock expose-fd listeners mode 666
    stats timeout 20m
```

Our hope was to be able to send out our string of commands, but we couldn't:
```
...
2022-12-13 20:07:03  debug: all good
2022-12-13 20:07:06  debug: agent socket signalled error
2022-12-13 20:07:06  debug: connecting agent: errno=111 Connection refused
2022-12-13 20:07:10  debug: connecting agent; errno=111 Connection refused
```

From our other tooling, we can confirm it's been quite a while since a reload (graceful), either earlier or after the degradation:
```
2022-12-13 19:33:40.592534542 --  args (reload, haproxy)
2022-12-13 19:33:40.593472853 --  reloading haproxy
...
2022-12-13 20:20:47.045221793 --  args: (reload, haproxy)
2022-12-13 20:20:47.046608401 --  reloading haproxy
```

We have our last telemetry data at exactly 20:07:00; error reported a few seconds later:
```
2022-12-13 20:07:04.909090548 /var/run/haproxy1.sock: errno 104 during send/recv: Connection reset by peer
2022-12-13 20:07:04.916581650 /var/run/haproxy2.sock: errno 104 during send/recv: Connection reset by peer
2022-12-13 20:07:04.916615463 /var/run/haproxy3.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916630340 /var/run/haproxy4.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916642874 /var/run/haproxy5.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916655111 /var/run/haproxy6.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916667067 /var/run/haproxy7.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916679290 /var/run/haproxy8.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916694330 /var/run/haproxy_backend10.sock: errno 111 during send/recv: Connection refused
2022-12-13 20:07:04.916706894 /var/run/haproxy_backend20.sock: errno 111 during send/recv: Connection refused
(...)
```

Our TX drop coincides with 20:07:04, when telemetry was lost.  So both TCP listeners (internet) and UNIX sockets (our internal traffic) are affected, and not promptly with regard to a reload.  The TX load at the time was very low, about 3.5 Gbps, so it doesn't appear to be triggered by either 1) a reload or 2) a specific network rate.

This particular instance is running:
```
HAProxy version 2.4.17-9f97155 2022/05/13 - https://haproxy.org/
```

Apart from losing telemetry, we still have no telling signs nor a pattern of when/why this happens.  This is by far our most elusive issue.

**Comment by rnsanchez** [comment]
2.6.7 also suffers from this:
```
HAProxy version 2.6.7-c55bfdb 2022/12/02
```

**Comment by capflam** [comment]
Sorry, the issue fell into oblivion. Is it still relevant ?

**Comment by rnsanchez** [comment]
This one is very peculiar.  I don't think we had it happening recently on 2.7.8 _but_ we're doing a lot of reloads (for different and healthy reasons, though).  However, we have no hypothesis as to what triggers it; it felt totally random whenever it happened.

If it helps closing the ticket, upon hitting it again we can update/reopen.

**Comment by capflam** [comment]
Thanks ! I'm closing the issue. It is not too hard to reopen it if necessary.
